devnet-sdk/controller/kt/mutate_env.go

```diff
@@ -0,0 +1,101 @@
+package kt
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/devnet-sdk/descriptors"
+
+	"github.com/ethereum-optimism/optimism/kurtosis-devnet/pkg/kurtosis/api/interfaces"
+)
+
+// a hack because some L2 services are duplicated across chains
+type redundantService []*descriptors.Service
+
+func (s redundantService) getEndpoints() descriptors.EndpointMap {
+	if len(s) == 0 {
+		return nil
+	}
+
+	return s[0].Endpoints
+}
+
+func (s redundantService) setEndpoints(endpoints descriptors.EndpointMap) {
+	for _, svc := range s {
+		svc.Endpoints = endpoints
+	}
+}
+
+func (s redundantService) refreshEndpoints(serviceCtx interfaces.ServiceContext) {
+	endpoints := s.getEndpoints()
+
+	publicPorts := serviceCtx.GetPublicPorts()
+	privatePorts := serviceCtx.GetPrivatePorts()
+
+	for name, info := range publicPorts {
+		endpoints[name].Port = int(info.GetNumber())
+	}
+	for name, info := range privatePorts {
+		endpoints[name].PrivatePort = int(info.GetNumber())
+	}
+
+	s.setEndpoints(endpoints)
+}
+
+func findSvcInEnv(env *descriptors.DevnetEnvironment, serviceName string) redundantService {
+	if svc := findSvcInChain(env.L1, serviceName); svc != nil {
+		return redundantService{svc}
+	}
+
+	var services redundantService = nil
+	for _, l2 := range env.L2 {
+		if svc := findSvcInChain(l2.Chain, serviceName); svc != nil {
```

Comment:
Ah I thought service names had to be unique within an enclave

Reply:
at the kurtosis level yes, but we currently have redundant entries in the descriptor (that map to the same kurtosis service)